genienlp/util.py

```diff
@@ -445,21 +448,25 @@ def elapsed_time(log):
 
 
 def make_data_loader(dataset, numericalizer, batch_size, device=None, paired=False, max_pairs=None, train=False,
-                     valid=False, append_question_to_context_too=False, override_question=None, override_context=None):
+                     append_question_to_context_too=False, override_question=None, override_context=None):
     
-    iterator = Iterator(dataset,
-                        batch_size,
-                        shuffle=train,
-                        repeat=train,
-                        use_data_batch_fn=train,
-                        use_data_sort_key=train)
+    all_features = NumericalizedExamples.from_examples(dataset, numericalizer, device=device,
+                                  paired=paired and train, max_pairs=max_pairs, groups=dataset.groups,
+                                  append_question_to_context_too=append_question_to_context_too,
+                                  override_question=override_question, override_context=override_context)
+    all_f = []
+    for i in trange(len(all_features.example_id), desc='Converting dataset to features'):
+        all_f.append(NumericalizedExamples(example_id=[all_features.example_id[i]],
+                            context=SequentialField(value=all_features.context.value[i][:all_features.context.length[i]], length=all_features.context.length[i], limited=all_features.context.limited[i][:all_features.context.length[i]]),
+                            question=SequentialField(value=all_features.question.value[i][:all_features.question.length[i]], length=all_features.question.length[i], limited=all_features.question.limited[i][:all_features.question.length[i]]),
+                            answer=SequentialField(value=all_features.answer.value[i][:all_features.answer.length[i]], length=all_features.answer.length[i], limited=all_features.answer.limited[i][:all_features.answer.length[i]]),
+                            decoder_vocab=all_features.decoder_vocab))
```

Comment:
Wait this code is suspicious. The first statement will do the actual tokenization, and the second statement is just creting the minibatches right? Do you create the minibatches only once at the beginning of the dataset?

Reply:
`NumericalizedExamples.from_examples(dataset ...` tokenizes the whole dataset to get DecoderVocab correctly. Then `all_f` breaks it down to individual examples (just reorders them in lists and removes padding).
Actual minibatches are created inside `NumericalizedExamples.collate_batches` which pads the examples again.

Reply:
Removing the pad and adding it later in `collate_batches` is a bit hacky, but has no effect on the speed since padding is really fast.

Reply:
We already do a pass over all examples in the Numericalizer to build the generatic vocab for the DecoderVocab, so now we have to do another one to build the other part of the DecoderVocab? This stuff is ripe for a cleanup. Not holding this PR for now, but definitely worth cleaning up.